File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n =
  if n <= 64 && n >=1
    then Just (2 ^ (n-1))
    else Nothing    

total :: Integer
total = sum $ map (\ i -> just (square i)) [1..64]

just :: Maybe a -> a
just (Just a) = a

File: solutions/julia/grains/1/grains.jl
"""Calculate the number of grains on square `square`."""
function on_square(square)
    check(square)
    BigInt(2)^(square-1)
end

"""Calculate the total number of grains after square `square`."""
function total_after(square)
    check(square)
    BigInt(2)^square - 1
end

function check(square)
    if square<1 || square>64
        throw(DomainError("wrong square"))
    end
end